src/uct/ib/base/ib_device.c

```diff
@@ -176,6 +194,89 @@ static void uct_ib_device_get_locality(const char *dev_name,
     *numa_node = (status == UCS_OK) ? n : -1;
 }
 
+static void
+uct_ib_device_async_event_dispatch(uct_ib_device_t *dev,
+                                   const uct_ib_async_event_t *event)
+{
+    uct_ib_async_event_val_t *entry;
+    khiter_t iter;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter = kh_get(uct_ib_async_event, &dev->async_events_hash, *event);
+    if (iter != kh_end(&dev->async_events_hash)) {
+        entry = &kh_value(&dev->async_events_hash, iter);
+        if (entry->cb) {
+            ucs_callbackq_add_safe(entry->cbq, entry->cb, entry->arg,
+                                   UCS_CALLBACKQ_FLAG_ONESHOT);
+        } else {
+            entry->flag = 1;
+        }
+    }
+    ucs_spin_unlock(&dev->async_event_lock);
+}
+
+ucs_status_t
+uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
```

Comment:
1. uct_ib_device_async_event_register (create entry with kh_put) - set cbq and resource_id
2. uct_ib_device_async_event_wait (type, res_id, cb, arg)